common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java

```diff
@@ -0,0 +1,261 @@
+
+        // Having thread sleep for 1 second for calling activity to receive the result from
+        // prepareForBrokerResumeRequest, thus the receiver for listening broker result return
+        // can be registered. openLinkInBrowser will launch activity for going to
+        // playstore and broker app download page which brought the calling activity down
+        // in the activity stack.
+        final int threadSleepForCallingActivity = 1000;
+        try {
+            Thread.sleep(threadSleepForCallingActivity);
+        } catch (InterruptedException e) {
+            Logger.verbose(TAG, "Error occurred when having thread sleeping for 1 second.");
+        }
+
+        HashMap<String, String> parameters = StringExtensions.getUrlParameters(url);
+        String link = parameters.get(INSTALL_URL_KEY)
+                .replace(AuthenticationConstants.Broker.BROWSER_EXT_PREFIX, "https://");
+        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(link));
+        getActivity().getApplicationContext().startActivity(intent);
+        view.stopLoading();
+
+        return true;
+    }
+
+    private boolean processInvalidUrl(@NonNull final WebView view, @NonNull final String url) {
+        if (isBrokerRequest(getActivity().getIntent())
+                && url.startsWith(AuthenticationConstants.Broker.REDIRECT_PREFIX)) {
+            Logger.error(TAG, "The RedirectUri is not as expected.", null);
+            Logger.errorPII(TAG, String.format("Received %s and expected %s", url, getRequest().getRedirectUri()), null);
+            returnError(ErrorStrings.DEVELOPER_REDIRECTURI_INVALID,
+                    String.format("The RedirectUri is not as expected. Received %s and expected %s", url,
+                            getRequest().getRedirectUri()));
+            view.stopLoading();
+            return true;
+        }
+
+        if (url.toLowerCase(Locale.US).equals("about:blank")) {
+            Logger.verbose(TAG, "It is an blank page request");
+            return true;
+        }
+
+        if (!url.toLowerCase(Locale.US).startsWith(AuthenticationConstants.Broker.REDIRECT_SSL_PREFIX)) {
+            Logger.error(TAG, "The webView was redirected to an unsafe URL.", null);
+            returnError(ErrorStrings.WEBVIEW_REDIRECTURL_NOT_SSL_PROTECTED, "The webView was redirected to an unsafe URL.");
+            view.stopLoading();
+            return true;
+        }
+
+        return false;
+    }
+
+    private void returnError(final String errorCode, final String errorMessage) {
+        Intent resultIntent = new Intent();
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_CODE, errorCode);
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE, errorMessage);
+
+        if (getRequest() != null) {
+            if (getRequest() instanceof AzureActiveDirectoryAuthorizationRequest) {
```

Comment:
Please correct me if I was wrong, since MicrosoftStsAuthorizationRequest extends AzureActiveDirectoryAuthorizationRequest, these 2 case could be merged into one.